### ISSUE TITLE
Get-ChildItem not throwing consistent errors for path not found

### ISSUE THREAD

**Issue by DarkLite1** [PowerShell/PowerShell#21141]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

When working with the Pester test suite we regularly use the default provided drive `TestDrive:\` described [here](https://pester.dev/docs/usage/testdrive). 

For one reason or another the error thrown by `Get-ChildItem` for a `Path not found` is different depending on the drive used.

```powershell
$params = @{
    # LiteralPath = 'c:\NotExisting' # Cannot find path 'C:\NotExisting' because it does not exist.
    LiteralPath = 'TestDrive:\NotExisting' # A parameter cannot be found that matches parameter name 'File'.
    File        = $true
    ErrorAction = 'Stop'
}
Get-ChildItem @params
```

### Expected behavior

```console
In all cases where the path is not found, we would expect `Get-ChildItem` to throw the error:

>  Cannot find path 'x:\xxx' because it does not exist.
```

### Actual behavior

```console
The actual behaviour is inconsistent depending on the drive.
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.1
PSEdition                      Core
GitCommitId                    7.4.1
OS                             Microsoft Windows 10.0.20348
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by mklement0** [comment]
I cannot reproduce this - neither on Windows nor on macOS.
I suspect this is somehow Pester-related (either to Pester itself or to your specific tests).
If you replace with `TestDrive:` with `Temp:` and run your `Get-ChildCommand` command stand-alone, do you get the expected error message?
(As an aside: It isn't great that the expected error message expresses the non-existent path as a _native_ path rather than in the terms of the specified _PS drive_.)

**Comment by 237dmitry** [comment]
> I cannot reproduce this

This error only occurs when the entire path does not exist:

```powershell
$ Get-ChildItem not:\exist -File         # -Path and -LiteralPath; -Directory and -File
Get-ChildItem: A parameter cannot be found that matches parameter name 'File'.
```

**Comment by jhoneill** [comment]
> > I cannot reproduce this
> 
> This error only occurs when the entire path does not exist:
> 
> ```powershell
> $ Get-ChildItem not:\exist -File         # -Path and -LiteralPath; -Directory and -File
> Get-ChildItem: A parameter cannot be found that matches parameter name 'File'.
> ```

Surely that is expected ?  -File and -Directory  are dynamic parameters which only apply if the path is a file system  
e.g.
```
> Get-ChildItem HKLM:\SOFTWARE\ -file
Get-ChildItem: A parameter cannot be found that matches parameter name 'file'.
```
If the drive name is not a file system (and that includes a non-existent drive) the error is thrown trying to bind the parameters, if it does exist but the path is non existent the parameters bind and the error comes from inside the cmdlet.

**Comment by 237dmitry** [comment]
Even if the disk does not exist, it would be more logical to see an error not about File or Directory, but about the fact that this path does not exist.

```
$ dir N:\ot\exist
Get-ChildItem: Cannot find drive. A drive with the name 'N' does not exist.

$ dir N:\ot\exist -File
Get-ChildItem: A parameter cannot be found that matches parameter name 'file'.
```

**Comment by mklement0** [comment]
Good point - it's about the non-existence of the drive, in which case no _provider context_ can be inferred.

Indeed, from a user's perspective it would be nicer if the error about the path not existing took precedence over the inability to bind then-unknown dynamic parameters - the current error, while technically justified, is confusing.

That said, implementing this without special-casing may or may not be possible.

_Hopefully_ - I'm not familiar with the innards of the parameter binder - the following is possible:

* If the `-Path` or `-LiteralPath` of a provider cmdlet is being bound...
* ... and the _provider context cannot be inferred_ due to a non-existent drive in an _absolute_ path being passed (in a _relative_ path, the provider context is _implied_, namely by the current location), report an error saying that no such drive exists.

The challenge is to prioritize this error among subsequent errors that may result from trying to bind to dynamic parameters of specific providers that are unavailable due to not recognizing _any_ provider, irrespective of the order in which arguments are passed.

That said, _tab-completion_ seems to have this order-independent analytic ability already:

```powershell
# Completes '-file' to '-File', because the subsequent $HOME value is recognized 
# as a file-system [provider] path.
TabExpansion2 "Get-ChildItem -file `"$HOME`"" -cursorColumn 17  | % completionmatches | % completiontext
'---'
# Does NOT complete '-file', because no provider context can be inferred 
# from non-existent drive NoSuch:
TabExpansion2 'Get-ChildItem -file NoSuch:' -cursorColumn 17  | % completionmatches | % completiontext
```

**Comment by jhoneill** [comment]
> Even if the disk does not exist, it would be more logical to see an error not about File or Directory

Yes, it would. But -File is a dynamic parameter. It doesn't exist unless some conditions are met.  
I have a vague memory that `Get-ChildItem`  finds parameters like `-File` and `Directory` by asking the provider indicated by the path what it should add. If that's so then even if there were a way for a cmdlet to jump in with it's own custom code for writing an error when invalid parameters are specified (which there isn't) GCI couldn't say "-File is only for file system provider paths" because -File isn't something it knows about either.  

The best which is likely to be practical is a suggestion like the "This file wasn't on the path but is in the current directory" one.

**Comment by mklement0** [comment]
To be clear:

* This is _not_ about _relative_ paths, because their provider context is _unambiguously implied_: They relate to the provider underlying the _current location_ (as reflected in `$PWD` and the output from `Get-Location`).

* This is purely about _absolute_ (full) paths _with drive specifications_ that refer to _non-existent_ drives.

As an aside:
  * On Unix-like platforms, `/`- and `\`-prefixed paths are special-cased and are _invariably_ assumed to be _file-system_ paths. 
    * In other words, loosely speaking: `/` (and `\`) on Unix-like platforms is the equivalent of `C:\` on Windows.
    * E.g., `Set-Location Env:; Get-ChildItem \` _on Windows_ lists all environment variables, whereas _on Unix_ it lists the child directories of the file-system root.

**Comment by 237dmitry** [comment]
> / (and \) on Unix-like platforms is the equivalent of C:\ on Windows.

Because `/` is PSDrive

What is `\`? Not alias...

**Comment by mklement0** [comment]
Yes, `/` is a PS drive, but it's special-cased in terms of syntax, otherwise, say, `Get-ChildItem /:tmp` should work, but it doesn't.

In seemingly all providers, `/` and `\` can be used _interchangeably_ as path separators (which is [problematic in the registry](https://github.com/PowerShell/PowerShell/issues/5536#issuecomment-346917545), for instance), and this is apparently also applied to `/` as a _drive_.

The upshot on Unix-like systems is that on any drive other than the file-system one with root `/` you cannot target such a drive's root directory with just `/` - you must use the drive spec; e.g., `Set-Location Temp:; Get-Item /` vs. `Set-Location Temp:; Get-Item Temp:/`

**Comment by 237dmitry** [comment]
> The upshot on Unix-like systems is that on any drive other than the file-system one with root / you cannot target such a drive's root directory with just / - you must use the drive spec

Yes, differences in file hierarchy between Linux and Windows force us to use different methods for the same operations. 
This difference is conditional, since drives in Windows can be mounted in directories and their letter designation removed, you get one root directory. That's why I assumed that the problem was with `/` PSDrive, in the name itself.

But I don't consider this a serious problem, I'm quite happy with using relative paths for the root directories of PSDrives.

**Comment by jdhitsolutions** [comment]
Added to the Cmdlet WG for review

**Comment by SteveL-MSFT** [comment]
@PowerShell/wg-powershell-cmdlets discussed this.  We agree that the current error message is misleading and will confuse users.  The root cause seems to be these [lines which catches the exception and simply returns that there are no dynamic parameters](https://github.com/PowerShell/PowerShell/blob/a6005f3879c5c3a344231f8d612604263aa45dd9/src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs#L254-L257).  The fix might simply be removing these lines and allowing the exception to be surfaced.  However, we are concerned since this code has been there forever if this might inadvertently break something we can't anticipate.  So we suggest anyone trying to fix this to wrap this as an Experimental Feature so we can assess through community feedback if it has unintended side effects

**Comment by CityguyUSA** [comment]
Get-ChildItem has vasts amounts of problems just returning valid results or trying to return results and then the interaction between itself and -ea has unpredictible results.  It's created a product that's not dependible.  This needs to be resolved or it may well end Powershell.  Everytime I run it something else happens.  The whole idea is not to have to babysit a script.

**Comment by jhoneill** [comment]
@CityguyUSA   you might need to be more specific.  Are things worse now than when Windows PowerShell 1.0 was launched ?because a lot of time has passed without this "ending PowerShell", and there aren't lots of open bugs about this.

**Comment by CityguyUSA** [comment]
Will you make the repairs?  Because I'm not going to waste my time documenting or re-documenting these problems.  As for Powershell who even verifies that it works correctly?  People assume that it's been verified by MS so very few people will even look.  When MS launches a product they should be guaranteeing it's a well tested product but because they have no liability they don't care.  I won't even go into the exhausting over wordy Powershell commands themselves.

**Comment by CityguyUSA** [comment]
If you fixed it an updated it this would have action wouldn't it?

**Comment by DarkLite1** [comment]
Has this been implemented?

**Comment by AikenBM** [comment]
This is still an issue on Powershell v7.6.6 on Server 2025 or Windows 11.

```
PS C:\> gci q:\
Get-ChildItem: Cannot find drive. A drive with the name 'q' does not exist.
PS C:\> gci q:\ -File
Get-ChildItem: A parameter cannot be found that matches parameter name 'File'.
PS C:\> gci q:\ -Directory
Get-ChildItem: A parameter cannot be found that matches parameter name 'Directory'.
````

**Comment by jhoneill** [comment]
> This is still an issue on Powershell v7.6.6 on Server 2025 or Windows 11.
> 
> ```
> PS C:\> gci q:\
> Get-ChildItem: Cannot find drive. A drive with the name 'q' does not exist.
> PS C:\> gci q:\ -File
> Get-ChildItem: A parameter cannot be found that matches parameter name 'File'.
> PS C:\> gci q:\ -Directory
> Get-ChildItem: A parameter cannot be found that matches parameter name 'Directory'.


that's expected and (sorry) correct behaviour and needs a significant change because parameter not found comes before the test for a valid drive. and those are dynamic parameters which only exist for a file system drive. 

Consider this 
```
>new-psdrive -name U -PSProvider Registry -Root HKEY_CURRENT_USER
> gci u: 
# returns registry data
gci u: -Directory   # note you can't tab expand Directory or file. 
Get-ChildItem: A parameter cannot be found that matches parameter name 'Directory'.
gci u: -file
Get-ChildItem: A parameter cannot be found that matches parameter name 'File'.
```
If the dynamic parameters don't exist - and they won't be called into being if U doesn't point to a file system, PowerShell does not even call the `Get-ChildItem `cmdlet, and it is the cmdlet which checks if the path is valid.  

`gci  Microsoft.PowerShell.Core\Registry::HKEY_CURRENT_USER ` is a valid path but not a file path, so no file or directory option
`gci u:` again valid but not a file path. 

If you don't like variable: , function: or alias: as names you can make drives named V, F or A.  Powershell can't assume a single letter name is a file system.

**Comment by CityguyUSA** [comment]
Makes the whole product questionable if it can't even traverse a directory structure.

**Comment by jhoneill** [comment]
> Makes the whole product questionable if it can't even traverse a directory structure.

"Because my comprehension of the product is so weak, not working how I think  it  makes it 'questionable' " ... really ? 

If you come from a cmd background drives are only file systems.  If you come from a unix background all manner of things are jammed into the file system

Anyone explaining PowerShell to new users explains `Get-PSDrive`  pretty early. "Oh look we have a drives for variables, and for functions and drives for the registry, and certificates, and if you these modules they create drives for more things".  
 So `gci  variable: ` gives you list of variables, and look it's not like a directory listing (and let's go off into PowerShell commands not *printing* output , but sending objects to something that does and while we are there, not having to parse their own command lines either)  
That's why commands are named New-Item, Remove-Item, Copy-Item, Rename-Item, and get-Item/ Get-ChildItem 
But then a bright student asks "but how does Get-ChildItem know how to traverse database objects like it does files?" which tees up more explanation of providers.   
And a bright student also asks "but what about telling Get-Childitem to only get something only available on one provider, And now we explain dynamic parameters  and the _exceptional_ student says "But hang on... that means I can write a command in a script in one place that requests directories in the root of a memory card as drive E: and if there is no drive E: when I run it, I get a bad parameter error without even calling because the dynamic parameter vanishes ?"    "Yup", "But I don't like that." "Then write wrappers or proxy functions which work the way you do like".  

Sure ideas that you have *Item commands which aren't confined to files and folders; and that drives aren't either; or that parameters might appear and disappear depending on context, or commands not parsing command lines or printing output, are too much for some people

**Comment by AikenBM** [comment]
@jhoneill I understand the argument. I understand that the system needs to decide on a provider before it can determine which provider-specific cmdlet to use. I get why it feels like a cart-before-the-horse problem.

But then how do you explain this:

```
PS U:\> gci \\nonexistantserver2112\share\ -File
Get-ChildItem: Cannot find path '\\nonexistantserver2112\share\' because it does not exist.
```

"Oh," you say, "that's because it's clearly a network path, and the provider handler is handling that differently."

Which is not the point. The point is that there absolutely is something looking at the path and deciding which provider it needs to use based on the URI alone. It's _that handler_ that is not throwing the appropriate exception message for the default cmdlet.

What should happen then is that the null or failover provider is specifically able to handle all custom parameters. The cmdlet definitely is passing something through and trying to run _a_ default command after failing to determine that provider. Otherwise, there would be nothing to emit the unknown named parameter error at all, and you'd get an error that the provider could not be determined for the path specified. That failover version of the cmdlet needs something like a `DyamicParam {}` block just as a place to black hole parameters it isn't going to reference. Because if the first thing that the cmdlet has to do is determine the provider based on the path, then the error handler for that needs to be able to respond with that exception. I don't care if the error handler got a parameter it didn't recognize. I want the actual error.

Again, I understand that this means that the system that determines the provider needs to pass a specific exception to the dynamic cmdlets, and then those cmdlets need to handle that exception. I see and understand why it's not a trivial change. But this breaks predictable behavior of a nearly ubiquitous mistake people make: typing a path wrong or without mapping a drive. And as you point out, the system handles an unlimited profile

Otherwise, I'd rather just have provider-specific commands because discoverability at the cost of functionality is not a fair trade. The dynamic provider design doesn't work if something as basic as a path error changes the exception from "invalid path" to "unknown parameter" which can potentially completely change exception trapping.

**Comment by surfingoldelephant** [comment]
> But then how do you explain this:

Because that path isn't qualified with a PS drive. PS considers it a [`PROVIDER-DIRECT`](https://github.com/PowerShell/PowerShell/blob/8c1e79829783d153181be987b75c81317c6aca82/src/System.Management.Automation/namespaces/LocationGlobber.cs#L4557-L4579) path rather than `DRIVE-QUALIFIED`. `PROVIDER-DIRECT` paths get resolved using [the current location's provider](https://github.com/PowerShell/PowerShell/blob/8c1e79829783d153181be987b75c81317c6aca82/src/System.Management.Automation/namespaces/LocationGlobber.cs#L1241-L1254).

Hence if the current location's provider is `FileSystem`, its dynamic `-File` parameter is found:

```powershell
cd $HOME # FileSystem
gci \\nonexistantserver2112\share\ -File
# Error: Cannot find path '\\nonexistantserver2112\share\' because it does not exist.
```

But if it's a different provider that doesn't expose `-File`:

```powershell
cd HKLM: # Registry
gci \\nonexistantserver2112\share\ -File
# Error: A parameter cannot be found that matches parameter name 'File'.
```

**Comment by jhoneill** [comment]
Better answer crossed with mine.

**Comment by CityguyUSA** [comment]
> > Makes the whole product questionable if it can't even traverse a directory structure.
> 
> "Because my comprehension of the product is so weak, not working how I think it makes it 'questionable' " ... really ?
> 
> If you come from a cmd background drives are only file systems. If you come from a unix background all manner of things are jammed into the file system
> 
> Anyone explaining PowerShell to new users explains `Get-PSDrive` pretty early. "Oh look we have a drives for variables, and for functions and drives for the registry, and certificates, and if you these modules they create drives for more things". So `gci variable: ` gives you list of variables, and look it's not like a directory listing (and let's go off into PowerShell commands not _printing_ output , but sending objects to something that does and while we are there, not having to parse their own command lines either) That's why commands are named New-Item, Remove-Item, Copy-Item, Rename-Item, and get-Item/ Get-ChildItem But then a bright student asks "but how does Get-ChildItem know how to traverse database objects like it does files?" which tees up more explanation of providers. And a bright student also asks "but what about telling Get-Childitem to only get something only available on one provider, And now we explain dynamic parameters and the _exceptional_ student says "But hang on... that means I can write a command in a script in one place that requests directories in the root of a memory card as drive E: and if there is no drive E: when I run it, I get a bad parameter error without even calling because the dynamic parameter vanishes ?" "Yup", "But I don't like that." "Then write wrappers or proxy functions which work the way you do like".
> 
> Sure ideas that you have *Item commands which aren't confined to files and folders; and that drives aren't either; or that parameters might appear and disappear depending on context, or commands not parsing command lines or printing output, are too much for some people

Regardless of all the goobleygook you wrote it doesn't solve the problem of getting stuck in an infinite loop trying to process the data in the results set.  It doesn't throw an error and it doesn't know it's still processing the same result over and over.  I was going to document the details but I know that no one fixes these problems and without a commitment to resolve these problems I'm not spending an hour documenting them to sit and be closed because it hit some random time period.

**Comment by jhoneill** [comment]
> >  it doesn't solve the problem of getting stuck in an infinite loop trying to process the data in the results set. It doesn't throw an error and it doesn't know it's still processing the same result over and over. I was going to document the details but I know that no one fixes these problems and without a commitment to resolve these problems I'm not spending an hour documenting them to sit and be closed because it hit some random time period.

This is an open source repo. You can contribute a fix, or find someone who thinks it is a problem who can. But the the very small amount of available resources Microsoft provided for PowerShell are very, very, unlikely to get directed to change something which has been same since before the product was named "PowerShell" , is by design, and is mostly considered to be a matter of user education. Rudeness about the product tends to be taken badly by those working on it, and (in my role as biter old curmudgeon) I'll pass on the advice to be _nice_ to those who have the power to fix your problem.  
If you can't put an hour in, don't expect anyone else to put their time in either. But if you find the time, I'll take a small bet that someone will show you how to make the problem go away - worst case wrapping GCI in a proxy function.
